liota/transports/mqtt.py

```diff
@@ -44,29 +46,112 @@ def on_subscribe(self, client, userdata, mid, granted_qos):
+                    raise ValueError("Error : Wrong client certificate path.")
+            else:
+                client_cert_available = False
+
+            # Validate client key file path
+            if self.gw_identity.keyfile:
+                if os.path.exists(self.gw_identity.keyfile):
+                    client_key_available = True
+                else:
+                    raise ValueError("Error : Wrong client key path.")
+            else:
+                client_key_available = False
+
+            # Multiple conditions for certificate validations
+                # 1. Both Client certificate and key file should be present
+                # 2. If both are not there proceed without client certificate and key
+                # 3. If client certificate is not there throw an error
+                # 4. If client key is not there throw an error
+            if client_cert_available and client_key_available:
+                log.debug("Certificates : ", self.gw_identity.cacert, self.gw_identity.certfile, self.gw_identity.keyfile)
+
+                self.client.tls_set(self.gw_identity.cacert, self.gw_identity.certfile, self.gw_identity.keyfile,
+                                    cert_reqs=getattr(ssl, self.TLSConf.cert_required),
+                                    tls_version=getattr(ssl, self.TLSConf.tls_version), ciphers=self.TLSConf.cipher)
+            elif not client_cert_available and not client_key_available:
+                self.client.tls_set(self.gw_identity.cacert,
+                                    cert_reqs=getattr(ssl, self.TLSConf.cert_required),
+                                    tls_version=getattr(ssl, self.TLSConf.tls_version), ciphers=self.TLSConf.cipher)
+            elif not client_cert_available and client_key_available:
+                raise ValueError("Error : Client key found, but client certificate not found")
+            else:
+                raise ValueError("Error : Client certificate found, but client key not found")
+
+            log.info("TLS support is set up.")
+
+        # Set up username-password
+        if self.enable_authentication:
+            if not self.gw_identity.username:
+                raise ValueError("Username not found")
+            elif not self.gw_identity.password:
+                raise ValueError("Password not found")
+            else:
+                self.client.username_pw_set(self.gw_identity.username, self.gw_identity.password)
+
+        # Set QoS parameters
+        if self.QoSConf.qoslevel:
+            if self.QoSConf.qoslevel>3:
+                raise ValueError("Error: Invalid QoS level")
+            elif self.QoSConf.qoslevel>0:
+                self.client.max_inflight_messages_set(self.QoSConf.inflight)
+                self.client.max_queued_messages_set(self.QoSConf.queue_size)
+                self.client.message_retry_set(self.QoSConf.retry)
+
 
+        # Connect with MQTT Broker
+        self.client.on_connect = self.on_connect
```

Comment:
Move on_connect binding to the init method where other callback methods are registered